src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+            }
+            if (success) {
+                boolean skipFinal = false;
+                try {
+                    try {
+                        child.onNext(value);
+                    } catch (Throwable t) {
+                        skipFinal = true;
+                        Exceptions.throwIfFatal(t);
+                        try {
+                            child.onError(t);
+                        } finally {
+                            unsubscribe();
+                        }
+                        return;
                     }
-                } else if (r > 0) {
-                    // drain what was requested
-                    long toEmit = r;
-                    for (int i = 0; i < toEmit; i++) {
-                        Object o = svq.poll();
-                        if (o == null) {
-                            break;
-                        } else {
-                            on.accept(actual, o);
-                            emittedWhileDraining++;
+                    producer.produced(1);
+                    request(1);
+                    // check if some state changed while emitting
+                    synchronized (this) {
+                        skipFinal = true;
+                        if (!missed) {
+                            emitting = false;
+                            return;
                         }
+                        missed = false;
                     }
-                    // decrement the number we emitted from outstanding requests
-                    MergeProducer.REQUESTED.getAndAdd(mergeProducer, -emittedWhileDraining);
-                }
-                return emittedWhileDraining;
-            }
-            return 0;
-        }
-
-        final Func1<InnerSubscriber<T>, Boolean> DRAIN_ACTION = new Func1<InnerSubscriber<T>, Boolean>() {
-
-            @Override
-            public Boolean call(InnerSubscriber<T> s) {
-                if (s.q != null) {
-                    long r = mergeProducer.requested;
-                    int emitted = s.drainQueue();
-                    if (emitted > 0) {
-                        s.requestMore(emitted);
+                } finally {
+                    if (!skipFinal) {
+                        synchronized (this) {
+                            emitting = false;
+                        }
                     }
-                    if (emitted == r) {
-                        // we emitted as many as were requested so stop the forEach loop
-                        return Boolean.FALSE;
+                }
+                // in case there was some activity, continue with the regular emission loop
+                emitLoop();
+            } else {
+                // either someone else is emitting or there was no available request
+                // queue up the scalar
+                RxRingBuffer svq = scalarQueue;
+                if (svq == null) {
+                    svq = RxRingBuffer.getSpscInstance();
```

Comment:
Am I correct in understanding that this is thread-safe since `tryEmitScalar` will only ever be called synchronously from one `onNext` at a time even though other inner subscribers could concurrently have the `emitting` flag?

It's confusing because the code for `emitting` needs to be synchronized, but this doesn't need to be as it is affected by only the outer Observable, not the concurrently subscribed inner Observables.

Perhaps some explanation in docs as to the thread-safety would be valuable to others reading the code? I had to flip back and forth in code several times to convince myself it was safe.